### PR TITLE
Fix implementation of compute_first_deposit_lp_

### PR DESCRIPTION
**Pull request by benluelo**
## Issue

The current implementation of `compute_first_deposit_lp_` was erroneously using the balance instead of the deposit. This PR removes the balance parameter from the function and refactors the input into a struct for clarity.

### CODE REVIEW

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 1]
```diff
@@ -465,23 +465,29 @@ pub fn compute_deposit_lp(
 ///
 /// https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#42-liquidity-provider-token-lpt-math-updates
 /// Equation 6
-pub fn compute_first_deposit_lp_<T: PerThing>(
-	pool_assets: &[(u128, u128, T)],
+pub fn compute_first_deposit_lp_<T, I>(
+	pool_assets: I,
 	_f: T,
```
Comment (benluelo) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:470]: Do we need this field if it's always 0? cc @vimukthi-git @PoisonPhang

Comment (PoisonPhang) [reply code/parachain/frame/composable-maths/src/dex/constant_product.rs:470]: I was under the impression that we planned to apply fees here in the future

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 2]
```diff
@@ -465,23 +465,29 @@ pub fn compute_deposit_lp(
...
-) -> ConstantProductAmmResult<ConstantProductAmmValueFeePair> {
+) -> ConstantProductAmmResult<ConstantProductAmmValueFeePair>
+where
+	T: PerThing,
+	I: IntoIterator<Item = ConstantProductFirstDepositInput<T>>,
+	I::IntoIter: ExactSizeIterator,
```
Comment (benluelo) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:475]: I also refactored this fn to be a bit more general, IMO this is a much cleaner interface.

**Review by PoisonPhang — APPROVED** [review]
LGTM
Good catch

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 3]
```diff
@@ -490,6 +496,11 @@ pub fn compute_first_deposit_lp_<T: PerThing>(
 	Ok(ConstantProductAmmValueFeePair { value: product.safe_to_fixed_point()?, fee: 0 })
 }
 
+pub struct ConstantProductFirstDepositInput<T> {
```
Comment (vimukthi-git) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:499]: Let's not introduce new types to keep consistency for the moment with all interfaces/functions in this file. For now simply removing the `balance` parameter works for first deposit use as the balance is always 0 for first deposit.

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 4]
```diff
@@ -465,23 +465,29 @@ pub fn compute_deposit_lp(
 ///
 /// https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#42-liquidity-provider-token-lpt-math-updates
 /// Equation 6
-pub fn compute_first_deposit_lp_<T: PerThing>(
-	pool_assets: &[(u128, u128, T)],
```
Comment (vimukthi-git) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:469]: Just removing the first `u128` param here would just solve the issue without a refactor as explained elsewhere.

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs (hunk 3 above)
Comment (benluelo) [reply code/parachain/frame/composable-maths/src/dex/constant_product.rs:499]: What consistency is being kept? Newtypes are used throughout this file (`ConstantProductAmmResult<_>`, `ConstantProductAmmValueFeePair`, `AssetAmount`, etc). I think using a newtype is actually _more_ consistent with what we currently have

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs (hunk 4 above)
Comment (benluelo) [reply code/parachain/frame/composable-maths/src/dex/constant_product.rs:469]: See https://github.com/ComposableFi/composable/pull/2572. I would still like to get these changes merged afterwards

Comment (vimukthi-git) [reply code/parachain/frame/composable-maths/src/dex/constant_product.rs:469]: Yes, we can consider better interface/structure for this file later after we finish with Pablo overall restructure.

**Pull request closed without merging**